skipcpio/skipcpio.c

```diff
@@ -60,59 +109,82 @@ int main(int argc, char **argv)
-                        buf[sizeof(buf) - 1] = 0;
-                        s = fread(buf, CPIO_ENDLEN, 2, f);
-                        if (s <= 0)
-                                break;
+                        filename_length = parse_hex_str(buf.h.c_namesize, sizeof(buf.h.c_namesize));
+                        pos = align_up(pos + sizeof(struct cpio_header) + filename_length, CPIO_ALIGNMENT);
 
-                        h = memmem(buf, sizeof(buf), CPIO_END, sizeof(CPIO_END));
-                        if (h) {
-                                pos = (h - buf) + pos + CPIO_ENDLEN;
+                        filesize = parse_hex_str(buf.h.c_filesize, sizeof(buf.h.c_filesize));
+                        pos = align_up(pos + filesize, CPIO_ALIGNMENT);
+
+                        if (filename_length == (CPIO_ENDLEN + 1) && strncmp(buf.filename, CPIO_END, CPIO_ENDLEN) == 0) {
                                 fseek(f, pos, SEEK_SET);
                                 break;
                         }
-                        pos += CPIO_ENDLEN;
+
+                        if (fseek(f, pos, SEEK_SET) != 0) {
+                                perror("fseek");
+                                exit(1);
+                        }
+
+                        if (fread(&buf, sizeof(buf), 1, f) != 1) {
+                                perror("fread");
+                                exit(1);
+                        }
+
+                        if (buf.h.c_magic[0] != '0' || buf.h.c_magic[1] != '7' ||
+                            buf.h.c_magic[2] != '0' || buf.h.c_magic[3] != '7' ||
+                            buf.h.c_magic[4] != '0' || buf.h.c_magic[5] != '1') {
+                                fprintf(stderr, "Corrupt CPIO archive!\n");
+                                exit(1);
+                        }
                 } while (!feof(f));
 
                 if (feof(f)) {
                         /* CPIO_END not found, just cat the whole file */
                         fseek(f, 0, SEEK_SET);
                 } else {
                         /* skip zeros */
-                        while (!feof(f)) {
+                        do {
+                                char copy_buffer[512];
                                 size_t i;
 
-                                buf[sizeof(buf) - 1] = 0;
-                                s = fread(buf, 1, sizeof(buf) - 1, f);
+                                s = fread(copy_buffer, 1, sizeof(copy_buffer) - 1, f);
                                 if (s <= 0)
                                         break;
 
-                                for (i = 0; (i < s) && (buf[i] == 0); i++) ;
+                                for (i = 0; (i < s) && (copy_buffer[i] == 0); i++) ;
 
-                                if (buf[i] != 0) {
+                                if (copy_buffer[i] != 0) {
                                         pos += i;
+                                        if (pos & 3) {
```

Comment:
IIUC, the zeros here aren't strictly part of a cpio archive, so probably shouldn't be subject to the 4 byte alignment requirement. `Documentation/driver-api/early-userspace/buffer-format.rst` indicates that kernel accepts `0 or more occurrences`of `("\0" | cpio_archive | cpio_gzip_archive)` which would imply that "\0" can be unaligned.

Reply:
removed...